twilight-gateway/src/shard.rs

```diff
@@ -726,11 +724,11 @@ impl Shard {
                         ConnectionStatus::FatallyClosed { close_code } => {
                             return Err(ReceiveMessageError::from_fatally_closed(close_code))
                         }
-                        _ => unreachable!(
-                            "stream ended because websocket is closed (received close frame sets \
-                            status to disconnected or fatally closed) or because it errored (which \
-                            also sets status to disconnected)"
-                        ),
+                        _ => {
+                            // Abnormal closure without close frame exchange.
+                            self.disconnect(CloseInitiator::None);
+                            self.reconnect(None, 0).await?;
```

Comment:
Actually, we do need to return something here to indicate upstream that the shard's closing. Again, let's leave this as is but remember to check back on this prior to a release. I plan on rewriting large parts of the internals of `Shard` (cancel safety) during which I'll fix this too.

Reply:
Let's get this merged into next very soon then so we can start refactoring the shard internals